Activate preformat block on android platform

marecar3 opened the pull request:
Fixes [Active preformat block on the Android platform](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1264)

AztecEditor-Android PR: https://github.com/wordpress-mobile/AztecEditor-Android/pull/869
Gutenberg PR: https://github.com/WordPress/gutenberg/pull/18777
WPAndroid PR: https://github.com/wordpress-mobile/WordPress-Android/pull/10868

To test:
 - Open the demo app
 - Add a Preformatted block to the content
 - Write around on it
 - Try to merge it with other blocks

Special case:

1. Create a draft Post on Gutenberg Web
2. Insert bellow content
```
<!-- wp:preformatted -->
<pre class="wp-block-preformatted">just a text<br>line2<br></pre>
<!-- /wp:preformatted -->
```
3. Open mobile application
4. Open drafted Post
5. Place the caret on the end of the Post
6. Go back (leave Editor)
7. Open draft Post again

Result: There shouldn't be a red screen: https://github.com/wordpress-mobile/AztecEditor-Android/pull/869#issuecomment-552864686

Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

SergioEstevao commented:
@mchowning bug number 2 is a [know bug](https://github.com/WordPress/gutenberg/issues/18476), and it was happening in iOS and the web too, it's now fixed in GB master by this [PR](https://github.com/WordPress/gutenberg/pull/18656)

marecar3 commented:
Changed target branch to `develop` as we don't want to hurry with this feature (it still has some active issues)

hypest commented:
> 5. Observe the cursor briefly jump, that a new line is added to the block, but the cursor remains at its current position.

This is probably caused by https://github.com/WordPress/gutenberg/blob/bb0529f59746e67dae47077bf91833e42a9081c5/packages/rich-text/src/component/index.native.js#L650. Aztec is trimming spaces in other cases but in the `<pre>` case it will not so, the JS code need to cater for that case too. @marecar3 can you have a look there?

> 6. Press Enter again
> 7. Observe that the previously added new line is removed and the cursor remains unmoved

I wasn't able to replicate that 🤔

marecar3 commented:
> > 1. Observe the cursor briefly jump, that a new line is added to the block, but the cursor remains at its current position.
> 
> This is probably caused by https://github.com/WordPress/gutenberg/blob/bb0529f59746e67dae47077bf91833e42a9081c5/packages/rich-text/src/component/index.native.js#L650. Aztec is trimming spaces in other cases but in the `<pre>` case it will not so, the JS code need to cater for that case too. @marecar3 can you have a look there?

Thanks a lot, @hypest, you opened my eyes, will let you know when I fix it.

> > 1. Press Enter again
> > 2. Observe that the previously added new line is removed and the cursor remains unmoved
> 
> I wasn't able to replicate that 🤔

Hey @hypest, I have fixed that issue.

hypest commented:
>  it's now fixed in GB master by this PR

Already [commented in the Gutenberg PR](https://github.com/WordPress/gutenberg/pull/18777#issuecomment-559494281) so, when that PR gets rebased the fix will be incorporated.

marecar3 commented:
> > 1. Observe the cursor briefly jump, that a new line is added to the block, but the cursor remains at its current position.
> 
> This is probably caused by https://github.com/WordPress/gutenberg/blob/bb0529f59746e67dae47077bf91833e42a9081c5/packages/rich-text/src/component/index.native.js#L650. Aztec is trimming spaces in other cases but in the `<pre>` case it will not so, the JS code need to cater for that case too. @marecar3 can you have a look there?
> 
> > 1. Press Enter again
> > 2. Observe that the previously added new line is removed and the cursor remains unmoved
> 
> I wasn't able to replicate that 🤔

Hey @hypest, I have fixed the problem with this change: https://github.com/WordPress/gutenberg/pull/18777/commits/2faec829bc9c157d7042b47a4499dcf5ff32156e

The main concern with that change is that I needed to remove `pre` and also `br` as regex match will be true if I just remove `pre`. Not sure if it's a safe to remove `br`?

marecar3 commented:
> 👋 @marecar3 ! This is looking pretty good. There are just a two issues I'm seeing with newline handling on Android.
> 
> Also, I think we need to update the WPAndroid branch to reference the latest gutenberg and Aztec code.
> 
> ## 1. Adding newline when last character on a line is a space causes problems with consecutive empty lines
> Steps:
> 
> 1. Add preformat block
> 2. Enter text on line _that ends with a space_
> 3. Press Enter and observe a new line as expected
> 4. Press Enter again
> 5. _Observe the cursor briefly jump, that a new line is added to the block, but the cursor remains at its current position._
> 6. Press Enter again
> 7. _Observe that the previously added new line is _removed_ and the cursor remains unmoved_
> 
> ![preformat-blocks-space-newline mp4](https://user-images.githubusercontent.com/4656348/69671061-4a258100-1063-11ea-8cb9-3dd7888fd715.gif)
> 
> ## 2. Merging blocks with newlines loses newlines
> Steps:
> 
> 1. Add preformat block
> 2. Add content spanning multiple lines
> 3. Create another preformat block
> 4. (Optional) Add content to the new block
> 5. Merge the new block back into the first block by pressing backspace at the beginning of the second block
> 6. _Observe that all newlines in the two blocks appear to have been replaced with spaces_
> 
> ![preformat-lose-newlines-on-merge mp4](https://user-images.githubusercontent.com/4656348/69671420-fff0cf80-1063-11ea-852b-3a44686f1c94.gif)

Hey @mchowning, you can do another round of testing, thanks! :)

hypest commented:
> Hey @hypest, I have fixed the problem with this change: WordPress/gutenberg@2faec82
>
> The main concern with that change is that I needed to remove pre and also br as regex match will be true if I just remove pre. Not sure if it's a safe to remove br?
I can't remember the case where we have html block with only br, but maybe I am wrong?

Without diving too much into this yet, I think that the `willTrimSpaces` pattern should match what Aztec will do when it will trim spaces. But, since in the PRE block we tell Aztec to not trim anything at all, it means that changing the regex is not exactly relevant. Instead, how about finding a way to shortcircuit willTrimSpaces() to return `false` if we know we are in the PRE case?

marecar3 commented:
> > Hey @hypest, I have fixed the problem with this change: [WordPress/gutenberg@2faec82](https://github.com/WordPress/gutenberg/commit/2faec82)
> > The main concern with that change is that I needed to remove pre and also br as regex match will be true if I just remove pre. Not sure if it's a safe to remove br?
> > I can't remember the case where we have html block with only br, but maybe I am wrong?
> 
> Without diving too much into this yet, I think that the `willTrimSpaces` pattern should match what Aztec will do when it will trim spaces. But, since in the PRE block we tell Aztec to not trim anything at all, it means that changing the regex is not exactly relevant. Instead, how about finding a way to shortcircuit willTrimSpaces() to return `false` if we know we are in the PRE case?

Hey @hypest, I have implemented another approach which you can find there: https://github.com/WordPress/gutenberg/pull/18777/commits/60dc8388f7f17977f86be35fbcb8be3eced28f80
let me know if you have some doubt about it, thanks.

hypest commented on react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecText.java:
```diff
@@ -23,11 +24,13 @@
 import com.facebook.react.views.textinput.ReactTextInputLocalData;
 import com.facebook.react.views.textinput.ScrollWatcher;
 
+import org.jetbrains.annotations.NotNull;
```
Hmm, any particular reason for using the JetBrains `NotNull` instead of AndroidX's `NonNull`?

Asking since I'd find it kinda more "natural" to to use the AndroidX's annotations unless they don't suit us well.

hypest replied:
Left an additional comment on the particular change: https://github.com/wordpress-mobile/gutenberg-mobile/pull/1615/commits/a665093164086ad2f7ce6bed83e349da7cccfd46#r354960106

hypest commented on react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecText.java:
```diff
@@ -23,11 +24,13 @@
 import org.wordpress.aztec.ITextFormat;
 import org.wordpress.aztec.plugins.IAztecPlugin;
 import org.wordpress.aztec.plugins.IToolbarButton;
+import org.wordpress.aztec.watchers.EndOfBufferMarkerAdder;
```
This is an unused import at this stage, right? Can we remove it or is it an indication that some code that uses is got lost?

marecar3 replied:
Yup, removed it.

hypest commented on react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecText.java:
```diff
@@ -525,11 +547,19 @@ public void onTextChanged(CharSequence s, int start, int before, int count) {
         }
     }
+
+    private boolean shouldRedirectTextChangeCalls() {
```
Nit: what do you think about naming this method as `shouldDelegateTextChangeCalls` to match the class' name?

marecar3 replied:
Yup, I agree.

hypest commented on react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecText.java:
```diff
@@ -525,11 +547,19 @@ public void onTextChanged(CharSequence s, int start, int before, int count) {
     }
+
+    private boolean shouldRedirectTextChangeCalls() {
+        return (!mIsSettingTextFromJS || isPreTag()) && mListeners != null;
```
The conditional has now started to be harder to read/follow. Can we split it up into documented individual ones that return early? WDYT?

For example:
```
        if (mListeners == null) {
            // No listeners so, no-one to delegate the calls to
            return false;
        }
```

marecar3 replied:
Good suggestion. I have split condition into multiple individuals.

marecar3 commented:
Hey @hypest, I think that I have addressed all issues so this one is ready for another iteration of review.

hypest commented on react-native-aztec/android/build.gradle:
```diff
@@ -12,7 +12,7 @@ buildscript {
         espressoVersion = '3.0.1'
 
-        aztecVersion = '03dd2332c1deb7c26dac82a7638121ca7b4ca0fd'
+        aztecVersion = '754086b517b1da25869102482048086ef928fb5b'
```
I notice that this commit includes 3 changes that are not actually related (the Aztec red bump, the `org.jetbrains.annotations.NotNull` change and the expansion of the conditional statement. No harm done but, I'd probably suggest keeping unrelated changes to separate commits. That helps with code review but most importantly, better documentation of the code changes. Our future selves usually like that isolation when investigating issues :)

marecar3 replied:
Good to know that :)

hypest commented on react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecText.java:
```diff
@@ -171,7 +170,7 @@ public boolean onTextContextMenuItem(int id) {
 
     @Override
-    public float getPreformatBackgroundAlpha(@NotNull TypedArray styles) {
+    public float getPreformatBackgroundAlpha(@Nullable TypedArray styles) {
```
Looking at the [corresponding Aztec code here](https://github.com/wordpress-mobile/AztecEditor-Android/pull/869/files#diff-7ca71cd1f009ab8438e0901205001b14R715) it looks like the parameter is not nullable (no `?` in the parameter type). Perhaps `@NonNull` is the appropriate annotation here?

marecar3 replied:
Hey @hypest, yes you were right in the first place. 
For some reason, I mixed those two annotations.
It's fixed now. Sorry about misunderstanding.

marecar3 commented:
> Made another pass Marko and only one comment needs attention, the one about the annotation.

👋 Thanks @hypest, I have fixed the comment which points to fix annotation.
You can do another iteration of testing.

hypest commented:
Hmm, the Android device tests consistently fail on the `rotate device and continue adding blocks` test. I've tried re-triggering them 3 times already 🤔

mchowning commented on react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecText.java:
```diff
@@ -525,11 +546,29 @@ public void onTextChanged(CharSequence s, int start, int before, int count) {
+    }
+
+    private boolean isPreTag() {
+        return !TextUtils.isEmpty(mTagName) && mTagName.equals(PRE_TAG);
```
This is just a nice-to-have, but I think this could be shortened to be a bit more clear by just reversing the equals check so you get a null and empty check for free since `PRE_TAG` is a constant:

```
return PRE_TAG.equals(mTagName);
```

marecar3 replied:
Good point. Fixed.
